### ISSUE TITLE
Some alerts cannot be read properly using a screenreader

### ISSUE THREAD

**Issue by zersiax** [Microsoft/vscode#5288]
- VSCode Version: insiders latest
- OS Version: Windows 10 build latest

Steps to Reproduce:

1. Start with a new copy of VS Code.
2. Open, or create, a file with the .cs (c#) extension.
3. An alert pops up, telling you you need to download an extension to get support for this or so. I got this from the docs.
4. Try to read the alert and fail :)

Expected behavior:
- The alert is read out loud or at least actionable using a screenreader.

Perceived result:
- The alert is not read, cannot be interacted with in any way I've been able to figure out and is not announced either. Unless you go merrily hopping through the accessibility tree you won't even see an alert is there.

Not sure if it concerns just this type of alert, I think this may be happening all over the place for this kind of control.

**Comment by bpasero** [comment]
@zersiax whenever an alert shows up in VS Code we set its text content to an aria-alert container. Testing with NVDA I can get the alert text to be read when it happens. What screen reader are you using?

**Comment by zersiax** [comment]
This is the strange part, it is hit and miss for me. Sometimes I hear
the alert, sometimes I don't. I use NVDA as well.
Even if the alert is heard though, are there actionable controls in
these alerts sometimes? A link or button one can click? If so, those
won't be clickable properly even if the role=alert gets fired...

**Comment by bpasero** [comment]
Yes, sometimes the alert does have controls embedded. Do you have a suggestion how to handle those in the alert? The actions are accessible with tab keyboard navigation and announce itself when they get focus.

Still, it seems to me there is a general issue of alerts sometimes not being read and sometimes they are? If so, this sounds like an issue with the screen reader and how it handles the ARIA ROLE alert.

We fill an element like this:

<div role="alert" aria-atomic="true">the alert message</div>

**Comment by bpasero** [comment]
One thing that is worth mentioning is that we seem to abuse the use of alert a little bit for workarounds for other accessibility reason. E.g. I see that suggestions from intellisense also end up in there. I wonder if the flood of alerts in the system cause the screen reader to throttle this.

@alexandrudima maybe you can comment how often the alert is being triggered from the editor?

**Comment by zersiax** [comment]
The suggestions for intelliSense being an alert was actually a bit of
a hack because of a chromium bug that in the last few days has seen a
few improvements, so we may be able to take those out soon.
When I was in Redmond last week I did a demo for some of the on-sight
folks and ran into the general issue of sometimes losing track of
where I am in the program. Does this toolbar belong to the explorer or
the editor? Is this treeview attached to toolbar A or toolbar B?  How
do I quickly go to the working files list?

This led me to the idea of making regions you can cycle between using
a hotkey. For example, using the f6 hotkey in Windows Explorer jumps
you from the address bar, to the treeview of folders, to the searchbox
etc.
I think this will be a very welcome addition to Code's accessibility
layer as well, and it would solve the alert problem too. You can just
jump to the region that holds the alert and get the keyboard focus to
go there. This way a user can read the alert whenever they want to
(the disadvantage of role-alert is that once it's spoken once you
can't really get it back unless you trigger it again), get to the
actionable controls and you make Code about 3 times easier to navigate
as well.
I realize this is a separate thing altogether but it does solve this
particular issue as well once implemented I think. Any thoughts?

**Comment by bpasero** [comment]
Well but how do you get aware of the fact that there is an alert?

**Comment by zersiax** [comment]
That is a trickier one. As I said, the alert only really shows up
once. I guess the only thing we can do if we go with the idea I just
came up with is notify the alert is there and get the user to manually
move to its region once they need it. The problem with that approach
is that a user might miss the notification and won't know the alert is
there ...I briefly considered a small 'blip' sound to let a user know
an alert has popped up but I'not sure how spammy that would be with
the amount of alerts we are currently using. Do any of you have ideas
for this one? :)

**Comment by bpasero** [comment]
Ok so the issue is that the alert is there but then another thing triggers another alert and so our alert gets lost. We could maybe introduce more than one alert container but I am not sure what this means.

Until then I suggest to not abuse the alert as we do it now. @alexandrudima what do you think?

**Comment by alexandrudima** [comment]
@zersiax I really like the idea of F6 as a quick way to navigate between regions. With F6, you could very quickly navigate and focus the message which would result in it being read again. I have noticed NVDA has a shortcut for reading the status bar (which unfortunately doesn't work with us as there is no aria role for a status bar), but maybe we could add a keyboard shortcut that quickly presents the high-level status of vscode. e.g. 
- "Explorer viewlet, two side-by-side editors, one error message"
- "Git viewlet, one diff editor opened"
- "Debug viewlet, in debug mode, stopped in fileA.js, one editor opened"
Thus, you could press Shift-F6 to get a quick overview of what's going on and F6 to navigate to each region.

If we go for such an approach, it means we need to rethink a bit how we do aria alerts, as they should not be a fire and forget sort of thing.

@bpasero From the original repro steps, suggestions do not appear when creating a new file so no aria alert calls occur from the suggest widget or from the editor. Have you instrumented the `vs/base/browser/ui/aria` -> `alert` method and found such calls with these steps?

**Comment by bpasero** [comment]
@zersiax trying to reproduce the issue and cannot, here are my steps:
* cmd+shift+p
* git commit

=> an info shows up and is read to me when I use NVDA.

Does this simple test case not work for you already?

**Comment by zersiax** [comment]
Strangely enough I can't seem to reproduce it now. If I miss the alert
it is still ineractable near the bottom of the accessibility tree.
What I do notice is that I press ctrl to shut up speech after I've
gone into the picker because it says 'multiline edit press alt+f1 if
you are using a screenreader". It would be good if that message
wouldn't come up every time you focus on the edit field since it will
trigger pretty much anyone who wants to be efficient to hit ctrl to
shut up speech, muting the alert. Does that make sense?

**Comment by alexandrudima** [comment]
The "multiline edit press alt+f1 if you are using a screenreader" is the aria label of the editor, it is not an aria alert. When you open a file the aria label contains text such as "findController.ts. Text file editor.Press Alt+F1 if you are using a screen reader."

Should we take out the "Press Alt+F1 if you are using a screenreader" part given it becomes so quickly so annoying?

**Comment by zersiax** [comment]
Well ...if I remember correctly doing this the first time toggles
screenreader features to true, right? Maybe only show it when
screenreader is false. This seems counter-intuitive but walk with me
here.
A user who is using a screenreader will want these features enabled,
and after that its very unlikely they will need the hotkey again since
it just tells you screenreader support is on.
So, when this flag is already toggled on there's no need to read out
that label. Does that make sense?

**Comment by alexandrudima** [comment]
@zersiax The entire Alt+F1 is probably not needed any more except to communicate that Ctrl+M can escape the Tab trapping in the editor. The experimental screen reader support is since 1.0 enabled by default.

**Comment by alexandrudima** [comment]
#5412 tracks that I should remove the out-of-date text

**Comment by zersiax** [comment]
Hi,

I think the control+m thing might be better stowed somewhere in a
documentation page, this is not something people would need to be
reminded about every time the editor gains focus I think. I liked your
suggestion about the statusbar thing, btw :)
I just did some checking and it doesn't seem to come up with this
message when you for example accept an autoComplete suggestion, thats
good at least :) Speaking of, that bug you ran into on that front
should be fixed in latest chromium :)

**Comment by bpasero** [comment]
@zersiax so you say the fact that you click back into the editor quickly after the alert pops up might have caused the screen reader to stop reading out the alert?

**Comment by zersiax** [comment]
I don't click sir :) What I was referring to is the ctrl hotkey, which
stops screenreader speech completely. Its often used if a screenreader
starts reading outa bit of text that you don't care about or have
already heard. You can just interrupt it and go on with your work,
more efficient that way. But in this case the inclusion of the whole
'press alt+f1" etc.  will make people press that hotkey if they
quickly want to resume working after having run something like git
commit which would make them not hear the alert as well.

**Comment by bpasero** [comment]
@zersiax understood. @alexandrudima so maybe we should only have the accessibility help in the editor once and afterwards we do not have it anymore so that it is not being read each time you focus the editor?

**Comment by bpasero** [comment]
@alexandrudima ping, should we show the Alt+F1 help only once the first time the user clicks into the editor?

**Comment by alexandrudima** [comment]
I can remove it completely, and handle its discoverability through documentation. Let's track its removal in #6908

**Comment by bpasero** [comment]
@alexandrudima nice!

**Comment by bpasero** [comment]
@zersiax I suggest you give our next release a try and see if #6908 improved this.
